File: c++/algorithm/Treap.cpp
#include <iostream>
#include <cstdlib>

using namespace std;


template <typename Key>
struct Node {

	Key key;
	int priority;
	int size;
	Node* left = nullptr;
	Node* right = nullptr;

	Node(const Key key) : key(key), priority(rand()), size(1){
	}

	void setLeft(Node* newLeft){
		left = newLeft++;
		calcSize();
	}

	void setRight(Node* newRight){
		right = newRight++;
		calcSize();
	}

	void calcSize() {
		size = 1;
		if(left) size += left->size;
		if(right) size += right->size;
	}
};

template <typename T>
using NodePair = pair<Node<T>*, Node<T>*>;

template <typename T>
NodePair<T> split(Node<T>* root, T key){

	if(root == nullptr) return {nullptr, nullptr};

	if(root->key < key){
		NodePair<T> right_side = split(root->right, key);
		root->setRight(right_side.first);
		return {root, right_side.second};
	}

	NodePair<T> left_side = split(root->left, key);
	root->setLeft(left_side.second);

	return {left_side.first, root};
}

template <typename T>
Node<T>* insert(Node<T>* root, Node<T>* node){
	if(root == nullptr) return node;

	if(root->priority < node->priority){
		NodePair<T> splitted = split(root, node->key);
		node->setLeft(splitted.first);
		node->setRight(splitted.second);
		return node;
	}

	if(node->key < root->key){
		root->setLeft(insert(root->left, node));
	}
	else {
		root->setRight(insert(root->right, node));
	}
	return root;
}

template <typename T>
Node<T>* merge (Node<T>* a, Node<T>* b){
	if (a == nullptr) return b;
	if (b == nullptr) return a;
	if (a->priority < b->priority){
		b->setLeft(merge(a, b->left));
		return b;
	}
	a->setLeft(merge(a->right, b));
	return a;
}

template<typename T>
Node<T>* erase(Node<T>* root, T key) {
	if(root == nullptr) return root;

	if(root->key == key){
		Node<T>* ret = merge(root->left, root->right);
		delete root;
		return ret;
	}

	if(key < root->key) 
		root->setLeft(erase(root->left, key));
	else
		root->setRight(erase(root->right, key));
	return root;
}

template <typename T> 
Node<T>* kth(Node<T>* root, int k){
	int left_size = 0;
	if(root->left != nullptr) left_size = root->left->size;
	if(left_size <= k) return kth(root->left, k);
	if(left_size + 1 == k) return root;
	return kth(root->right, k);
}

template <typename T>
Node<T>* countLessThan(Node<T>* root, T key){
	if(root == nullptr) return 0;
	if(root->key >= key) return countLessThan(root->left, key);
	int left_side = (root->left) ? root->left->size : 0;
	return left_side + 1 + countLessThan(root->right, key);
}

int main(void){


	return 0;
}

File: python/algorithm/Treap.py
import random


class Node:
    def __init__(self, value, left=None, right=None):
        self.value = value
        self._left = left
        self._right = right
        self.priority = random.randrange(0, 100)
        self.size = 0
        pass

    @property
    def left(self):
        return self._left

    @left.setter
    def left(self, left):
        self._left = left
        self.calcSize()

    @property
    def right(self):
        return self._right

    @right.setter
    def right(self, right):
        self._right = right
        self.calcSize()

    def calcSize(self):
        size = 1
        if self.left:
            size += self.left.size
        if self.right:
            size += self.right.size
        self.size = size

    @property
    def min(self):
        if not self._left:
            return value
        return self.left.min

    @property
    def max(self):
        if not self.right:
            return value
        return self.right.max

    def __str__(self):
        return f'Node({self.value})'


def split(root: Node, key):
    if not root:
        return None, None

    if root.value < key:
        left, right = split(root.right, key)
        root.right = left
        return root, right

    left, right = split(root.left, key)
    root.left = right
    return left, root


def insert(root: Node, node: Node):

    if not root:
        return node

    if root.priority < node.priority:
        first, second = split(root, node.value)
        node.left = first
        node.right = second
        return node

    if root.value > node.value:
        root.left = insert(root.left, node)
    else:
        root.right = insert(root.right, node)
    return root


def merge(a: Node, b: Node):
    assert a.max < b.min

    if not a:
        return b
    if not b:
        return a

    if a.priority < b.priority:
        b.left = merge(a, b.left)
        return b

    a.right = merge(a.right, b)
    return a


def erase(root: Node, key):
    if not root:
        return root

    if root.value == key:
        ret = merge(root.left, root.right)
        del root
        return ret

    if root.value < key:
        erase(root.left, key)
    else:
        erase(root.right, key)

    return root


def kth(root: Node, k: int):
    if not root:
        return None

    left_size = 0
    if root.left:
        left_size = root.left.size
    if k <= left_size:
        return kth(root.left, k)
    if k == left_size + 1:
        return root

    return kth(root.right, k - left_size - 1)


def count_less_than(root: Node, key):
    if not root:
        return 0
    if root.value >= key:
        return count_less_than(root.left, key)

    left_side = 0
    if root.left:
        left_side = root.left.size
    return left_side + 1 + count_less_than(root.right, key)


if __name__ == "__main__":
    random.seed(43)

    root = Node(0)
    for i in range(1, 11):
        root = insert(root, Node(i))

    for i in range(1, 12):
        print(kth(root, i))

    print(count_less_than(root, 10))

    pass
